Fix gemma2 tokenizer convert

ngxson opened the pull request:
Ref comment:
- Investigation: https://github.com/ggerganov/llama.cpp/issues/8240#issuecomment-2200674401
- Fix proposal: https://github.com/ggerganov/llama.cpp/issues/8240#issuecomment-2200778010

The output model is capable of tokenizing special tokens (used in chat templates):

```
test.txt

<start_of_turn>user
hello discards HOW are you<end_of_turn>
```

```
$ ./llama-tokenize -m ggml-model-f16.gguf -f test.txt

...

     2 -> '<bos>'
   106 -> '<start_of_turn>'
  1645 -> 'user'
   108 -> '
'
 17534 -> 'hello'
  9027 -> ' disc'
  2050 -> 'ards'
 31874 -> ' HOW'
   708 -> ' are'
   692 -> ' you'
   107 -> '<end_of_turn>'
   108 -> '
'
```

Perplexity is also improved from `8.9711` to `7.8952` (I'm using q8_0 because colab notebook does not have enough VRAM for f16)

```
$ !./llama.cpp/llama-perplexity -f ./llama.cpp/wikitext-2-raw/wiki.test.raw -ngl 99 -m ./gemma2/ggml-model-q8_0.gguf -c 1024

[277]7.8267,[278]7.8213,[279]7.8372,[280]7.8424,[281]7.8534,[282]7.8573,[283]7.8760,[284]7.8952,
Final estimate: PPL = 7.8952 +/- 0.05648
```

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

ngxson committed:
fix gemma2 tokenizer convert

ngxson committed:
remove scores

abetlen commented:
@ngxson great job, thank you for the fix. It looks like this now matches the huggingface tokenizer on all of the examples I've run it through.

Probably a nit but it might be better to have a seperate `_create_vocab_sentencepiece` method that's used by `set_vocab` as well as `_set_vocab_sentencepiece` instead of having an optional return based on a keyword arg.

abetlen commented:
@ngxson this seems closer to the correct tokenization however when testing this with the cli / server I've noticed that newlines generation seems broken.

ngxson commented:
@abetlen Thanks for testing that. I've just tried on my side, and I can confirm that the new line is also broken. The model outputs token ID 235248 for new line, which is mapped to `"▁": 235248,` on `tokenizer.json`. That makes no sense, probably the original model is broken.

ngxson commented:
Edit: sorry I made a mistake. The output new line token is correct (token ID 108), I'm investigating this further

ngxson committed:
improve code, fix new line issue

ngxson commented:
@abetlen Turns out new line and all tokens after ID 108 are marked as control, while they should be normal token. I fixed my code and it should work correctly now:

```
> list 10 fruits
Here are 10 fruits:

1. Apple
2. Banana
3. Orange
4. Strawberry
5. Grapefruit
6. Mango
7. Pineapple
8. Watermelon
9. Blueberry
10. Raspberry
```

Tokenized (main.log):

```
'<start_of_turn>':106, 'user':1645, '':108, 'list':1701, ' ':235248, '1':235274, '0':235276, ' fruits':16803, '<end_of_turn>':107, '':108, '<start_of_turn>':106, 'model':2516, '':108, 'Here':4858, ' are':708, ' ':235248, '1':235274
```

I also split the code into `_create_vocab_sentencepiece` and `_set_vocab_sentencepiece`. This way the function is easier to be reuse.

bartowski1182 commented:
this seems like the correct one. It even properly tokenizes the prompt with discards noted in the discussion as disc and ards

@ggerganov if you want to take a look

eran-medan commented on convert-hf-to-gguf.py:
```diff
@@ -2345,7 +2350,19 @@ class Gemma2Model(Model):
+        for i in range(108):
+            # including <unusedX>, <start_of_turn>, <end_of_turn>
+            toktypes[i] = SentencePieceTokenTypes.CONTROL
+        self.gguf_writer.add_tokenizer_model("llama")
```
I know it’s merged, and it’s a nitpick, and ignores the [rule of 3](https://en.m.wikipedia.org/wiki/Rule_of_three_(computer_programming)) but especially for someone with little understanding of what the code is actually doing (it’s all magic to me) it would benefit from a separate method for this and the sequence of calls lines 582-586 

while removing a small duplication, it also can serve as a helper to understand what it’s doing. And someday someone may fix something in once place and not the other.

❤️

ngxson replied:
I think you’re new to the code base.. if you have a look on the other parts of the file, there’re even more duplications. Not because we don’t care about this, but sometimes duplications make it more visible what the code does.
